### ISSUE TITLE
A non-obtrusive convention for documenting packages?

### ISSUE THREAD

**Issue by Aerlinger** [mesosphere/universe#165]
As others have noted it can be difficult to mentally parse each of the different configurations, versions, schemas, etc. for all of these different packages. Personally, I've noticed everything tends to blur together when navigating between package folders -- a problem that will obviously get worse as the repo grows.

One solution I'd like to propose is adopting the convention of adding a README.md scoped to either the package or each package version. This README would include human readable documentation along with a description, a manifest of all the configuration options, examples, and so on, perhaps similar to the information as shown in the DCOS docs (https://docs.mesosphere.com/services/spark/), but more concise. 

Consider it like a "man page", that doesn't depend on a Unix environment.

Here's an example:

```
└── foo
    ├── 0
    │   ├── command.json
    │   ├── config.json
    │   ├── marathon.json
    │   ├── package.json
    │   └── README.md
    ├── 1
    │   ├── command.json
    │   ├── config.json
    │   ├── marathon.json
    │   ├── package.json
    │   └── README.md
    └── ...
```

The README could be optionally scoped to the package folder level as well, giving precedence to the README local to that version where present:

```
└── foo
    ├── README.md
    ├── 0
    │   ├── command.json
    │   ├── config.json
    │   ├── marathon.json
    │   ├── package.json
    │   └── README.md
    ├── 1
    │   ├── command.json
    │   ├── config.json
    │   ├── marathon.json
    │   └── package.json
    └── ...
```

*Benefits:*

- Github treats the README.md differently than other files, rendering formatted markdown in the browser when navigating a repo.
- "OCP" friendly: extends rather than modifies existing functionality, and doesn't require any initial burden of code implementation or modification.
- Markdown renders to HTML so this could be embedded in the online documentation, rather than requiring unnecessary duplication.
- Enables simple and accessible documentation from the command line in the form of something like `dcos package spark --help` or `dcos package --info spark`. This sort of documentation seems to be currently lacking within the `dcos package` subcommand (I always seem to be going back and forth between my terminal and several tabs of DCOS documentation in my browser). 

Anyhow, just wanted to throw this out there and hear any comments/feedback/criticisms.

**Comment by ConnorDoyle** [comment]
:+1: this sounds great.

**Comment by pyronicide** [comment]
This sounds pretty fantastic to me. Could we integrate the pre/post install notes into the README so they're all in one place?

**Comment by ssk2** [comment]
Great suggestion!

**Comment by jdef** [comment]
Does this proposal include auto-generating the README.md based on
information already contained in the .json files? For example the pre/post
notes (as suggested by Thomas) but also docs for all of the config options,
etc. I wouldn't want to have to keep redundant sets of documentation up to
date.

**Comment by jsancio** [comment]
cc @tsenart

**Comment by tsenart** [comment]
JSON is already human readable. If we want more context on what the JSON means, let's define [schemas](http://json-schema.org/) and create tools to leverage them.

**Comment by Aerlinger** [comment]
@tsenart I don't agree that JSON is readable in the sense that it can also double as a form of documentation, however, I definitely do agree that introducing JSON schemas would be the way to go. At the very least there should be a way of providing documentation on the "public" configuration options available to an end user of the CLI that wouldn't require knowledge of the internals or implementation details of the package itself. Unless I'm mistaken, nothing exists for this purpose currently. Any user who wants to manually configure a package has to grok the JSON files and how each attribute/setting affects the package itself, which is often an exercise in trial and error.

@jdef Keeping documentation consistent is definitely important, and it would make sense to have the options of generating a boilerplate readme from the json schema. However, one issue to consider would be preventing overwriting any "hand made" documentation so to speak.

If it would be useful, I'd be happy to submit a PR to show a more concrete implementation example.

**Comment by BenWhitehead** [comment]
@tsenart and @Aerlinger There are already json schemas for each of the different document types see here: https://github.com/mesosphere/universe/tree/version-1.x/repo/meta/schema

With regard to exposing the configuration options that are available for a given package, I believe this is something the CLI time is currently working on surfacing so you aren't having to dig through package files.  I well defer to @jsancio on details.

**Comment by jsancio** [comment]
Hey @Aerlinger. We are working on improving the CLI on this regard. @mgummelt is working on providing an option to the CLI to discover all of the configuration parameters of a package.
